src/libpriv/rpmostree-core.c

```diff
@@ -670,6 +690,227 @@ on_hifstate_percentage_changed (DnfState   *hifstate,
+  pkg_meta_dict = g_variant_dict_new (pkg_meta);
+
+  header = _rpmostree_vardict_lookup_value_required (pkg_meta_dict,
+                                                     "rpmostree.metadata",
+                                                     (GVariantType*)"ay",
+                                                     error);
+  if (!header)
+    {
+      g_autofree char *nevra = cache_branch_to_nevra (cachebranch);
+      g_prefix_error (error, "In commit %s of %s: ", cached_rev, nevra);
+      return FALSE;
+    }
+
+  *out_header = g_steal_pointer (&header);
+  return TRUE;
+}
+
+static gboolean
+checkout_pkg_metadata_by_dnfpkg (RpmOstreeContext *self,
+                                 DnfPackage       *pkg,
+                                 GCancellable     *cancellable,
+                                 GError          **error)
+{
+  const char *nevra = dnf_package_get_nevra (pkg);
+  g_autofree char *cachebranch = rpmostree_get_cache_branch_pkg (pkg);
+  g_autoptr(GVariant) header = NULL;
+
+  if (!get_header_variant (self->pkgcache_repo, cachebranch, &header,
+                           cancellable, error))
+    return FALSE;
+
+  return checkout_pkg_metadata (self, nevra, header, cancellable, error);
+}
+
+gboolean
+rpmostree_pkgcache_find_pkg_header (OstreeRepo    *pkgcache,
+                                    const char    *nevra,
+                                    const char    *expected_sha256,
+                                    GVariant     **out_header,
+                                    GCancellable  *cancellable,
+                                    GError       **error)
+{
+  g_autoptr(GVariant) header = NULL;
+  g_autoptr(GHashTable) refs = NULL;
+  GHashTableIter it;
+  gpointer k;
+
+  /* there's no safe way to convert a nevra string to its
+   * cache branch, so let's just do a dumb lookup */
+
+  if (!ostree_repo_list_refs_ext (pkgcache, "rpmostree/pkg", &refs,
+                                  OSTREE_REPO_LIST_REFS_EXT_NONE, cancellable,
+                                  error))
+    return FALSE;
+
+  g_hash_table_iter_init (&it, refs);
+  while (g_hash_table_iter_next (&it, &k, NULL))
+    {
+      const char *ref = k;
+      g_autofree char *cache_nevra = cache_branch_to_nevra (ref);
+
+      if (g_str_equal (nevra, cache_nevra))
+        {
+          if (expected_sha256 != NULL)
+            {
+              g_autofree char *commit_csum = NULL;
+              g_autoptr(GVariant) commit = NULL;
+              g_autofree char *actual_sha256 = NULL;
+
+              if (!ostree_repo_resolve_rev (pkgcache, ref, TRUE,
+                                            &commit_csum, error))
+                return FALSE;
+
+              if (!ostree_repo_load_commit (pkgcache, commit_csum,
+                                            &commit, NULL, error))
+                return FALSE;
+
+              if (!get_commit_header_sha256 (commit, &actual_sha256, error))
```

Comment:
Hmm, I don't *think* so. We still want to error out, since we can't verify that it's the same package we want. We should be fine from the backwards compatibility perspective too, since we would only request a match if we've previously been asked to install a local RPM (and we always re-unpack them), so the header would get created.